File: sdks/src/ts/decompiler.ts
import { BlobReader, ZipReader, Uint8ArrayWriter } from "@zip.js/zip.js";
interface Node {
    id: string;
    type: string;
    talker: number;
    message: string;
    feeling: number;
    assetId: number;
    outPoints: { label: string; nextId: string }[];
}
interface Character {
    name: string;
    feelings: Record<number, number>;
}
interface Noun {
    refer: string;
    calls: string[];
}
interface Asset {
    type: string;
    data: string | ArrayBuffer;
}
interface ProjectData {
    nodes: Node[];
    characters: Character[];
    feelings: string[];
    nouns: Noun[];
    assets: Asset[];
    entryNode: string | null;
}
function base64ToArrayBuffer(base64: string) {
    const binaryString = atob(base64);
    const len = binaryString.length;
    const buffer = new ArrayBuffer(len);
    const uint8Array = new Uint8Array(buffer);
    for (let i = 0; i < len; i++) {
        uint8Array[i] = binaryString.charCodeAt(i);
    }
    return buffer;
}
async function readBlob(file: Blob): Promise<string> {
    return new Promise((resolve) => {
        const reader = new FileReader();
        reader.addEventListener("load", () => {
            if (typeof reader.result !== "string") return;
            const buffer = reader.result;
            resolve(buffer);
        });
        reader.readAsText(file);
    });
}
export class ScriptPlayer {
    public static allowedPairs = ["$;", "[]", "{}"];
    public static allowedSeparators = ":.>%~#@→↣↝↠↣↦⇀⇏⇒⇥⇨⇢⇰⇸⇻⇾▸▹▶▷►▻";
    public static nounSpliter = new RegExp(`[${ScriptPlayer.allowedSeparators.split("").map(char => "\\" + char).join("")}]`, "g");
    public static centerd = `\\w+${ScriptPlayer.nounSpliter.source}\\s*\\d+\\s*`;
    public static nounMatcher = new RegExp(ScriptPlayer.allowedPairs.map(pair => `(\\${pair[0]}${ScriptPlayer.centerd}\\${pair[1]})`).join("|"), "g");
    public static unknownNounTip = "▸未知名词◂";
    private project?: ProjectData;
    public format<T extends keyof Node>(node: Node, key: T): Node[T] {
        if (!this.project) throw new Error("Project not loaded");
        if (key === "message") {
            return node.message.replace(ScriptPlayer.nounMatcher, match => {
                if (!this.project) throw new Error("Project not loaded");
                const data = match.slice(1, -1).split(ScriptPlayer.nounSpliter);
                const [refer, index] = [data[0].trim(), Number(data[1]) - 1];
                if (!this.project.nouns.some(nounSrc => nounSrc.refer === refer)) return match;
                return this.project.nouns.find(nounSrc => nounSrc.refer === refer)?.calls[index] ?? ScriptPlayer.unknownNounTip;
            }) as Node[T];
        } else if (key === "talker") {
            const character = this.project.characters[node.talker];
            if (!character) return node.talker as Node[T];
            return (character.name ?? node.talker) as Node[T];
        } else if (key === "feeling") {
            return (this.project.feelings[node.feeling] ?? node.feeling) as Node[T];
        } else if (key === "assetId") {
            const asset = this.project.assets[node.assetId]
            if (!asset) return node.assetId as Node[T];
            if (asset.type === "script") return asset.data as Node[T];
            else return node.assetId as Node[T];
        }
        return node[key];
    }
    public async open(file: Blob, password?: string): Promise<ProjectData> {
        this.project = await ScriptPlayer.decompile(file, password);
        return this.project;
    }
    public findNodeById(nodeId: string): Node {
        if (!this.project) throw new Error("Project not loaded");
        const result = this.project.nodes.find(node => node.id === nodeId);
        if (!result) throw new Error(`Node not found: ${nodeId}`);
        return result;
    }
    public async play(callback: (node: Node) => number | void | Promise<number | void>, nodeId?: string): Promise<number> {
        if (!this.project) throw new Error("Project not loaded");
        const nodeID = nodeId ?? this.project.entryNode;
        if (!nodeID) throw new Error("No entry node found, please set a entry target");
        const node = this.findNodeById(nodeID);
        const state = await callback(node) ?? -1;
        if (state >= 0) {
            const point = node.outPoints[state];
            if (!point) return state;
            const { nextId } = point;
            if (!nextId || nextId === "null") return state;
            return await this.play(callback, nextId);
        } else {
            return state;
        }
    }
    public static async decompile(file: Blob, password?: string): Promise<ProjectData> {
        const projectData: ProjectData = {
            nodes: [],
            characters: [],
            feelings: [],
            nouns: [],
            assets: [],
            entryNode: null
        };
        try {
            const decodedBytes = base64ToArrayBuffer(await readBlob(file));
            file = new Blob([decodedBytes]);
        } catch { }
        const zipReader = new ZipReader(new BlobReader(file), { password });
        const entries = await zipReader.getEntries();
        for (const entry of entries) {
            if (entry.directory || !entry.getData) continue;
            const { buffer: content } = await entry.getData<Uint8Array>(new Uint8ArrayWriter(), { password }) as { buffer: ArrayBuffer };
            let decodedContent: string | ArrayBuffer;
            try {
                decodedContent = new TextDecoder().decode(content);
            } catch {
                decodedContent = content;
            }
            if (entry.filename.endsWith(".node")) {
                const lines = decodedContent.toString().split("\n");
                const node: Node = {
                    id: entry.filename.replace(".node", "").replace(".entry", ""),
                    type: lines[0],
                    talker: Number(lines[1]),
                    message: lines[2],
                    feeling: Number(lines[3]),
                    assetId: Number(lines[4]),
                    outPoints: lines[5]
                        .split(",")
                        .filter(Boolean)
                        .map(pair => {
                            const [label, nextId] = pair.split(":");
                            return { label, nextId };
                        }),
                };
                projectData.nodes.push(node);
                if (entry.filename.endsWith(".entry.node")) {
                    projectData.entryNode = node.id;
                }
            } else if (entry.filename.endsWith(".character")) {
                const lines = decodedContent.toString().split("\n");
                projectData.characters.push({
                    name: lines[0],
                    feelings: Object.fromEntries(
                        lines[1]
                            .split(",")
                            .filter(Boolean)
                            .map(pair => pair.split(":").map(Number))
                    ),
                });
            } else if (entry.filename.endsWith(".feeling")) {
                projectData.feelings.push(decodedContent.toString().trim());
            } else if (entry.filename.endsWith(".noun")) {
                projectData.nouns.push({
                    refer: entry.filename.replace(".noun", ""),
                    calls: decodedContent.toString().split("\n"),
                });
            } else if (entry.filename.endsWith(".image") || entry.filename.endsWith(".video") || entry.filename.endsWith(".script")) {
                const assetType = entry.filename.split(".").pop()!;
                projectData.assets.push({
                    type: assetType,
                    data: decodedContent,
                });
            }
        }
        await zipReader.close();
        return projectData;
    }
}

File: sdks/src/python/decompiler.py
import pyzipper, io
import base64
import re

allowed_pairs = ["$;", "[]", "{}"]
allowed_separators = ":.>%~#@→↣↝↠↣↦⇀⇏⇒⇥⇨⇢⇰⇸⇻⇾▸▹▶▷►▻-"
noun_spliter = re.compile(f"[{re.escape(allowed_separators)}]")
centerd = f"\\w+{noun_spliter.pattern}\\s*\\d+\\s*"
noun_matcher = re.compile(
    "|".join(
        f"(\{pair[0]}{centerd}\{pair[1]})"
        for pair in allowed_pairs
    )
)
unknown_noun_tip = "▸未知名词◂"

class ScriptPlayer:
    def __init__(self):
        self.project_data = None

    @staticmethod
    def decompile(file_path: str | io.BytesIO | bytes, password: str = None):
        """
        Decompile the given file and extract assets.
        """
        # 检查是否是 Base64 编码
        if isinstance(file_path, (str, bytes)):
            try:
                if isinstance(file_path, str):
                    file_path = file_path.encode("utf8")
                decoded_bytes = base64.b64decode(file_path, validate=True)
                file_path = io.BytesIO(decoded_bytes)
            except (base64.binascii.Error, ValueError):
                pass  # 不是 Base64 编码，继续使用原始输入

        project_data = {
            "nodes": [],
            "characters": [],
            "feelings": [],
            "nouns": [],
            "assets": [],
        }
        byte_password = password.encode("utf8") if password else None
        with pyzipper.AESZipFile(file_path, "r") as zf:
            for file_name in zf.namelist():
                file_name: str
                with zf.open(file_name, "r", byte_password) as file:
                    content = file.read()
                    try:
                        content = content.decode("utf8")
                    except:
                        pass
                    if file_name.endswith(".node"):
                        lines = content.split("\n")
                        node = {
                            "id": file_name.strip(".node").strip(".entry"),
                            "type": lines[0],
                            "talker": lines[1],
                            "message": lines[2],
                            "feeling": lines[3],
                            "assetId": lines[4],
                            "outPoints": [
                                {
                                    "label": pair.split(":")[0],
                                    "nextId": pair.split(":")[1],
                                }
                                for pair in lines[5].split(",")
                                if pair
                            ],
                        }
                        project_data["nodes"].append(node)
                        if file_name.endswith(".entry.node"):
                            project_data["entryNode"] = node["id"]
                    elif file_name.endswith(".character"):
                        lines = content.split("\n")
                        character = {
                            "name": lines[0],
                            "feelings": {
                                int(k): int(v)
                                for k, v in (
                                    pair.split(":")
                                    for pair in lines[1].split(",")
                                    if pair
                                )
                            },
                        }
                        project_data["characters"].append(character)
                    elif file_name.endswith(".feeling"):
                        project_data["feelings"].append(content.strip())
                    elif file_name.endswith(".noun"):
                        noun = {
                            "refer": file_name.replace(".noun", ""),
                            "calls": content.split("\n"),
                        }
                        project_data["nouns"].append(noun)
                    elif file_name.endswith((".image", ".video", ".script")):
                        asset_type = file_name.split(".")[-1]
                        project_data["assets"].append(
                            {
                                "type": asset_type,
                                "data": content,
                            }
                        )
        return project_data

    def open(self, file_path, password: str = None):
        self.project_data = ScriptPlayer.decompile(file_path, password)
        return self.project_data

    def find_node_by_id(self, node_id: str):
        if not self.project_data:
            raise ValueError("Project not loaded")
        node = next((n for n in self.project_data["nodes"] if n["id"] == node_id), None)
        if not node:
            raise ValueError(f"Node with ID {node_id} not found")
        return node

    def play(self, callback, node_id: str = None):
        if not self.project_data:
            raise ValueError("Project not loaded")
        current_node_id = node_id or self.project_data.get("entryNode")
        if not current_node_id:
            raise ValueError("No entry node found, please set a target")
        while current_node_id:
            node = self.find_node_by_id(current_node_id)
            next_index = callback(node)
            if next_index < 0 or next_index >= len(node["outPoints"]):
                break
            current_node_id = node["outPoints"][next_index]["nextId"]

    def format(self, node: dict, key: str):
        if not self.project_data:
            raise ValueError("Project not loaded")
        if key == "message":
            return re.sub(
                noun_matcher,
                lambda match: self._replace_noun(match.group()),
                node["message"],
            )
        elif key == "talker":
            character = self.project_data["characters"][node["talker"]]
            return character["name"] if character else node["talker"]
        elif key == "feeling":
            return self.project_data["feelings"][node["feeling"]]
        elif key == "assetId":
            asset = self.project_data["assets"][node["assetId"]]
            return asset["data"] if asset and asset["type"] == "script" else node["assetId"]
        return node[key]

    def _replace_noun(self, match: str):
        data = re.split(noun_spliter, match[1:-1])
        refer, index = data[0].strip(), int(data[1]) - 1
        noun = next((n for n in self.project_data["nouns"] if n["refer"] == refer), None)
        if not noun or index < 0 or index >= len(noun["calls"]):
            return match
        return noun["calls"][index]
